File: leetcode/problems/language-specific/go/155.min-stack.go
/*
 * @lc app=leetcode id=155 lang=golang
 *
 * [155] Min Stack
 */

// @lc code=start
type MinStack struct {
	RealStack []int
	MinStack  []int
}

func Constructor() MinStack {
	return MinStack{}
}

func (this *MinStack) Push(val int) {
	this.RealStack = append(this.RealStack, val)
	if len(this.MinStack) <= 0 || val <= this.MinStack[len(this.MinStack)-1] {
		this.MinStack = append(this.MinStack, val)
	}
}

func (this *MinStack) Pop() {
	val := this.RealStack[len(this.RealStack)-1]
	this.RealStack = this.RealStack[:len(this.RealStack)-1]
	if val == this.MinStack[len(this.MinStack)-1] {
		this.MinStack = this.MinStack[:len(this.MinStack)-1]
	}
}

func (this *MinStack) Top() int {
	return this.RealStack[len(this.RealStack)-1]
}

func (this *MinStack) GetMin() int {
	return this.MinStack[len(this.MinStack)-1]
}

/**
 * Your MinStack object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Push(val);
 * obj.Pop();
 * param_3 := obj.Top();
 * param_4 := obj.GetMin();
 */
// @lc code=end

File: leetcode/problems/sessions/2023-05-01_2023-??-??/155.min-stack.py
#
# @lc app=leetcode id=155 lang=python3
#
# [155] Min Stack
#

# @lc code=start

# * Two Stacks Solution | O(1) Time | O(n) Space
# * n -> The number of push operations


class MinStack:
    def __init__(self):
        self.real_stack = []
        self.min_stack = []

    def push(self, val: int) -> None:
        self.real_stack.append(val)
        if not self.min_stack or val <= self.min_stack[-1]:
            self.min_stack.append(val)

    def pop(self) -> None:
        if self.real_stack.pop() == self.min_stack[-1]:
            self.min_stack.pop()

    def top(self) -> int:
        return self.real_stack[-1]

    def getMin(self) -> int:
        return self.min_stack[-1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()
# @lc code=end
